lib/opal/cache.rb

```diff
@@ -29,79 +29,147 @@ def fetch(*)
       end
     end
 
-    class FileCache
-      def initialize(dir: nil)
-        @root = File.expand_path '..', Opal.gem_dir
-        @dir = dir || find_dir
-      end
+    unless RUBY_ENGINE == 'opal'
+      class FileCache
+        def initialize(dir: nil, max_size: nil)
+          @dir = dir || self.class.find_dir
+          @root = File.expand_path('..', Opal.gem_dir)
+          # Store at most 32MB of cache - de facto this 32MB is larger,
+          # as we don't account for inode size for instance. In fact, it's
+          # about 50M. Also we run this check before anything runs, so things
+          # may go up to 64M or even larger.
+          @max_size = max_size || 32 * 1024 * 1024
+
+          tidy_up_cache
+        end
 
-      def fetch(*key)
-        key = key.join('/')
-        file = cache_filename_for(key)
+        def fetch(*key)
+          key = key.join('/')
+          file = cache_filename_for(key)
+
+          if File.exist?(file)
+            load_data(file)
+          else
+            compiler = yield
+            store_data(file, compiler)
+            compiler
+          end
+        end
 
-        if File.exist?(file)
-          load_data(file)
-        else
-          compiler = yield
-          store_data(file, compiler)
-          compiler
+        private def store_data(file, data)
+          out = Marshal.dump(data)
+          out = Zlib.gzip(out, level: 9)
+          File.binwrite(file, out)
         end
-      end
 
-      private def store_data(file, data)
-        out = Marshal.dump(data)
-        out = Zlib.gzip(out, level: 9)
-        File.binwrite(file, out)
-      end
+        private def load_data(file)
+          FileUtils.touch(file)
+          out = File.binread(file)
+          out = Zlib.gunzip(out)
+          Marshal.load(out) # rubocop:disable Security/MarshalLoad
+        end
 
-      private def load_data(file)
-        FileUtils.touch(file)
-        out = File.binread(file)
-        out = Zlib.gunzip(out)
-        Marshal.load(out) # rubocop:disable Security/MarshalLoad
-      end
+        # Remove cache entries that overflow our cache limit... and which
+        # were used least recently.
+        private def tidy_up_cache
+          entries = Dir[@dir + '/*.rbm.gz']
+
+          size_sum = entries.map { |i| File.size(i) }.sum
+          return unless size_sum > @max_size
+
+          # First, we try to get the oldest files first.
+          entries = entries.sort_by { |i| File.mtime(i) }
+          # Then, what's more important, is that we try to get the least
+          # recently used files first. Filesystems with relatime or noatime
+          # will get this wrong, but it doesn't matter that much, because
+          # the previous sort got things "maybe right".
+          entries = entries.sort_by { |i| File.atime(i) }
```

Comment:
Can these two be combined in a single cycle with `entries.sort_by! { |i| [File.mtime(i), File.atime(i)] }`?